### ISSUE TITLE
 converted core ml model doen't include classes in metadata

### ISSUE THREAD

**Issue by dhgokul** [ultralytics/yolov5#3238]
converted yolo5s.pt model to core ml model using cmd

`!python models/export.py --weights  weights/yolov5s.pt --img 640 --batch 1 --train `

in previous tried cmd
`!python models/export.py --weights  weights/yolov5s.pt --img 640 --batch 1 ` but coreml model fails to convert

once the core ml model  tried in ios below result got
![MicrosoftTeams-image (6)](https://user-images.githubusercontent.com/5689804/118823383-93603500-b8d6-11eb-942d-155af005ae50.png)

here no class found.

Tried in mac x code

**Comment by dhgokul** [comment]
not able to convert yolo5s model  to core ml model . as suggessted command
` !python models/export.py --weights weights/yolov5s.pt --img 640 --batch 1 `

tried both linux cpu machine  and in google colab. In both cases not showing the exact error on conversion
![Screenshot from 2021-05-20 18-29-39](https://user-images.githubusercontent.com/5689804/118982880-7a6c8800-b999-11eb-92c7-d56442c71d33.png)

so just removed exception in coreml conversion part in model.py file , got the error as follows in google colab
![Screenshot from 2021-05-20 18-32-03](https://user-images.githubusercontent.com/5689804/118983180-c7e8f500-b999-11eb-9255-181a55b2aad8.png)

just installed all in the requirement.txt

how to over come the issue.

**Comment by glenn-jocher** [comment]
@dhgokul good news 😃! Your original issue may now been fixed ✅ in PR #3262. This PR improves error handling for CoreML exports, displaying an error message that notifies the user that CoreML exports should be done when the model is in `.train()` mode with `python models/export.py --train`. To receive this update you can:
- `git pull` from within your `yolov5/` directory
- `git clone https://github.com/ultralytics/yolov5` again
- Force-reload [PyTorch Hub](https://pytorch.org/hub/ultralytics_yolov5/): `model = torch.hub.load('ultralytics/yolov5', 'yolov5s', force_reload=True)`
- View our updated notebooks:  <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>

Thank you for spotting this issue and informing us of the problem. Please let us know if this update resolves the issue for you, and feel free to inform us of any other **issues** you discover or **feature requests** that come to mind. Happy trainings with YOLOv5 🚀!

**Comment by dhgokul** [comment]
thanks for the quick response 

we just pulled the repo and  tried conversion

`!python models/export.py --weights weights/yolov5s.pt --img 640 --batch 1 --train`
the generated core model is tested on x code, but we have not any class labels, generally in other core ml mode list in metadata
![Screenshot from 2021-05-20 21-08-55](https://user-images.githubusercontent.com/5689804/119008271-a561d680-b9af-11eb-8e69-c5f9ccdb19aa.png)

here is the converted ml model of yolov5s.pt
[yolov5s_3.zip](https://github.com/ultralytics/yolov5/files/6517168/yolov5s_3.zip)
we tried this ml model only in ios x code.

**Comment by yohandarosa** [comment]
anyone had find a solution to include class labels in the CoreML model ?

**Comment by tylercollins590** [comment]
Was this ever solved? I’m having the same issue

**Comment by jiwon-hae** [comment]
Same issue here... Was this solved?

**Comment by Sic-1** [comment]
Still seems to be an issue. Any solutions?

**Comment by MaxNazarov93** [comment]
I figured out this problem and what you need is create ClassifierConfig and pass it into convert function

```
from coremltools.converters import ClassifierConfig

classifier_config = ClassifierConfig(class_labels=['class1', 'class2', 'class3', 'class4'])
ct_model = ct.convert(ts, inputs=[ct.ImageType('image', shape=im.shape, scale=1 / 255, bias=[0, 0, 0])], classifier_config=classifier_config)
```

**Comment by dhgokul** [comment]
Thanks @MaxNazarov93  will check it. Anyone try this and post in this group 

Thanks in advance!

**Comment by zjsjayce** [comment]
I also meet this problem

**Comment by BasanthVerma** [comment]
> I figured out this problem and what you need is create ClassifierConfig and pass it into convert function
> 
> from coremltools.converters import ClassifierConfig
> 
> classifier_config = ClassifierConfig(class_labels=['class1', 'class2', 'class3', 'class4'])
> ct_model = ct.convert(ts, inputs=[ct.ImageType('image', shape=im.shape, scale=1 / 255, bias=[0, 0, 0])], classifier_config=classifier_config)

@MaxNazarov93  this process adds the classes to the model and we can see the classes in the metadata. But when we run the model, we get:
`Error Domain=com.apple.vis Code=3 "The VNCoreMLTransform request failed" UserInfo={NSLocalizedDescription=The VNCoreMLTransform request failed, NSUnderlyingError=0x280a81590 {Error Domain=com.apple.CoreML Code=0 "The size of the output layer 'var_887' in the neural network does not match the number of classes in the classifier." UserInfo={NSLocalizedDescription=The size of the output layer 'var_887' in the neural network does not match the number of classes in the classifier.}}}`

We tried multiple Yolov5 that we had custom trained, we made sure that class_labels has the exact number of labels that we used for training. But we still keep getting this error.

So we ended up following the script from https://rockyshikoku.medium.com/convert-yolov5-to-coreml-also-add-a-decode-layer-113408b7a848 **_and it worked!_**
